### PR TITLE
Ring test, correct image center, enable write PSF object to pickle, record flux zero point in output exposure

### PR DESCRIPTION
**Pull request by mr-superonion**
1. Enable ring test to reduce error on additive bias (4x 45 degree pairs);
2. Enable dumping PSF object to disk (for reuse);
3. record magnitude zero point in the output exposures;
4. The center of the image was at ndim/2-1 for an image with indexing 1,2..ndim (that is the Gaism indexing). Now fix it to ndim/2+1. For example, a length 3 array, with indexing 1,2,3, the center should be at 2=(3+1)/2.

### CODE REVIEW

**Comment by esheldon** [comment]
Note rather than ring test we use the noise cancelling method of Pujols. Note in the mdet-lsst-sim repo I use both a +g1 and -g1 and store both in the output files

**Comment by mr-superonion** [comment]
Yes! I think it is better to use both together! (and this is normally what I did) since ring test is very efficient in reducing error on additive bias.

File: descwl_shear_sims/sim.py
Comment (beckermr) [inline descwl_shear_sims/sim.py:336]: Galsim is 1-offset by defaulkt but the Rubin science pipelines use 0-offset coordinates. Given how the center is used below I am pretty sure 0-offset is correct here.

File: descwl_shear_sims/sim.py
Comment (beckermr) [inline descwl_shear_sims/sim.py:741]:
```suggestion
    cost = np.cos(theta)
    sint = np.sin(theta)
    x2 = cost*x - sint*y
    y2 = sint*x + cost*y
```

**Review by beckermr — COMMENTED** [review]
Thanks for the PR! A few initial comments. I'll need to read more later too.

File: descwl_shear_sims/sim.py
Comment (mr-superonion) [reply descwl_shear_sims/sim.py:336]: Let's look into the details of the original code:
https://github.com/LSSTDESC/descwl-shear-sims/blob/5a693fd926f5e99ea1a0f76857dc66085e313ad5/descwl_shear_sims/sim.py#L322-L324
`cen` defined here adopts DM convention!
 
However, you can see the `cen` is first passed into Galsim and used to defined Galsim WCS:

https://github.com/LSSTDESC/descwl-shear-sims/blob/5a693fd926f5e99ea1a0f76857dc66085e313ad5/descwl_shear_sims/sim.py#L336-L342

After that, the Galsim WCS is used to define a DM WCS 
https://github.com/LSSTDESC/descwl-shear-sims/blob/5a693fd926f5e99ea1a0f76857dc66085e313ad5/descwl_shear_sims/sim.py#L395

and in the function here:
https://github.com/LSSTDESC/descwl-shear-sims/blob/5a693fd926f5e99ea1a0f76857dc66085e313ad5/descwl_shear_sims/wcs/dmwcs.py#L25-L29
The Galsim center is transformed to the DM center by doing ` Point2D(crpix[0]-1, crpix[1]-1)`. 
The problem is that at the beginning, the Galsim center does not use Galsim convention..

Comment (mr-superonion) [reply descwl_shear_sims/sim.py:336]: Let's assume the `dim=33`, then we have `cen=16`
The Gaism image should have grids at 1,2,3...33. 
The correct Galsim Center should be at (17,17), however we will get an inconsistent one which is (16,16) 
Then, if we pass this inconsistent Galsim WCS to the function `make_dm_wcs()`, we will get 
`stack_crpix=Point2D(15,15)`, which is not correct.

Comment (esheldon) [reply descwl_shear_sims/sim.py:336]: @rmjarvis Do what do you think of this, did we mess up the galsim WCS?

Comment (rmjarvis) [reply descwl_shear_sims/sim.py:336]: I don't know.  I would write a unit test that checks that the connection between some particular pixel and its Ra/Dec is right.  I'm not following enough of what you are doing here to know whether you did everything right.  But presumably making a test like that should be straightforward.

Comment (mr-superonion) [reply descwl_shear_sims/sim.py:336]: OK!! I will add a really simple test for one galaxy and force the offset to [0,0] to show that for the original code, the galaxy is not at the center of the image but has a (-1,-1) offset to the image center. I have done a similar test (but more complicated) in my shear estimation repo.. but I will put that one here..

Comment (beckermr) [reply descwl_shear_sims/sim.py:336]: We have a test like this where we coadd psfs. Maybe we don't check the offset there though.

**Comment by esheldon** [comment]
We need to run a full sim test to show this didn't introduce any shear biases.

**Comment by esheldon** [comment]
Please don't lint with black.  It makes pointless changes (e.g. single to double quotes) and creates conflicts with the flake8 we use for CI

**Comment by mr-superonion** [comment]
> Please don't lint with black. It makes pointless changes (e.g. single to double quotes) and creates conflicts with the flake8 we use for CI

wow, what is the software this repo used for auto-linting? Please let me what is the standard?

**Comment by esheldon** [comment]
I don't auto-format my code, I run flake8 within the editor and make the changes by hand.

**Comment by mr-superonion** [comment]
> We need to run a full sim test to show this didn't introduce any shear biases.

Yes!!! Let me make a test for shear estimation with 4 noiseless galaxies (45 degree rotated; with center offset; but fix the center of the measurement to the center of distortion/ rotation)

**Comment by esheldon** [comment]
This might be covered by the CI with shear_meas_tests subdir

**Comment by mr-superonion** [comment]
Note, the shear estimation test also make sure that the distortion/ rotation center is located at the center of the image array.

**Comment by beckermr** [comment]
What is biased?

**Comment by mr-superonion** [comment]
shear is biased if the measurement center is not the same as the center of the distortion/ rotation for this test!

**Comment by beckermr** [comment]
We've done tons and tons of simulations without this PR and found no shear bias. Under what conditions are you talking about? Are you not detecting objects?

**Comment by mr-superonion** [comment]
This is a special case, since I am using only 4 noiseless galaxies for ring test (each galaxy has an offset and then rotated by 45 degree wrt image center) and run metacal (no detection is run). 
This is to make sure that using 4 galaxies successfully removes shape noise.
Since no detection is run, and each galaxy has center offset from the image center, in the measurement we must force the center of the measurement to the center of the rotation, otherwise we will find a bias but this is because that the ring test does not fully remove shape noise.

File: descwl_shear_sims/shifts.py
Comment (esheldon) [inline descwl_shear_sims/shifts.py:218]: Actually the original doc string was quite wrong, this is  not gridded.

It should be something like "get a set of random shifts in a square"

File: descwl_shear_sims/shifts.py
Comment (esheldon) [inline descwl_shear_sims/shifts.py:239]: Is the idea that the user might use a buffer that is too  large?

Comment (esheldon) [reply descwl_shear_sims/shifts.py:239]: Maybe we should actually raise an exception if `dim - 2*buff < 0`

File: descwl_shear_sims/shifts.py
Comment (esheldon) [inline descwl_shear_sims/shifts.py:257]: not gridded right?  Maybe "get a set of random shifts on a disk"?

File: descwl_shear_sims/shifts.py
Comment (esheldon) [inline descwl_shear_sims/shifts.py:256]: It fills a disk, not a circle right?

File: descwl_shear_sims/sim.py
Comment (esheldon) [inline descwl_shear_sims/sim.py:423]: What does this do?

File: descwl_shear_sims/sim.py
Comment (esheldon) [inline descwl_shear_sims/sim.py:121]: Should be "The rotation angle of the objects, in units of radians" rather than exposure

File: descwl_shear_sims/sim.py
Comment (esheldon) [inline descwl_shear_sims/sim.py:319]: Probably should use this same doc string above

File: descwl_shear_sims/sim.py
Comment (mr-superonion) [reply descwl_shear_sims/sim.py:423]: https://hsc.mtk.nao.ac.jp/pipedoc/pipedoc_4_e/e_tool/index.html
See the last row of  table 1 in the link above.
In HSC pipeline, (I do not know why) the photometric calibration file is called frc...

File: descwl_shear_sims/sim.py
Comment (mr-superonion) [reply descwl_shear_sims/sim.py]: done!

File: descwl_shear_sims/shifts.py
Comment (mr-superonion) [reply descwl_shear_sims/shifts.py]: Done!!

File: descwl_shear_sims/shifts.py
Comment (mr-superonion) [reply descwl_shear_sims/shifts.py]: Done!!

File: descwl_shear_sims/shifts.py
Comment (mr-superonion) [reply descwl_shear_sims/shifts.py]: Done

File: descwl_shear_sims/sim.py
Comment (esheldon) [reply descwl_shear_sims/sim.py:424]: I mean, what do these operations do?  Do they set some metadata in the exposure?  How is it retrieved etc.

Comment (esheldon) [reply descwl_shear_sims/sim.py:424]: Thanks.  Do you mind putting that as a comment in the code for future reference?

Comment (mr-superonion) [reply descwl_shear_sims/sim.py]: Done!!

**Comment by mr-superonion** [comment]
> Where do we stand on this PR?

Sorry I have been busy with some other projects. 
I am waiting for the review from you and Matthew!!

**Review by esheldon — APPROVED** [review]

**Review by beckermr — APPROVED** [review]
There is an API change here which may need to be propagated into downstream users of the code.

**Comment by esheldon** [comment]
can you point that out?

File: descwl_shear_sims/galaxies.py
Comment (beckermr) [inline descwl_shear_sims/galaxies.py:589]: Here @esheldon. If you are trying to do a new `random_ring` simulation, you need to remember to pass this keyword here in addition to the other places `layout` is used.

**Comment by beckermr** [comment]
Right. Not a breaking change but could be broken in some cases.

**Comment by esheldon** [comment]
@mr-superonion can you pass on that keyword indicated above?

**Comment by mr-superonion** [comment]
I think now it is consistent!!

File: descwl_shear_sims/galaxies.py
Comment (esheldon) [inline descwl_shear_sims/galaxies.py:78]: In the past the default has been random, so we should also default to that here rather than making it a required argument

**Review by esheldon — CHANGES_REQUESTED** [review]

**Review by esheldon — APPROVED** [review]

**Comment by esheldon** [comment]
Thanks for this PR, I think this will be very useful.

**Pull request merged**
